src/Compilers/Core/Portable/PEWriter/MetadataVisitor.cs

```diff
@@ -80,11 +80,23 @@ public virtual void Visit(IEventDefinition eventDefinition)
             this.Visit(eventDefinition.GetType(Context));
         }
 
-        public void Visit(IEnumerable<IFieldDefinition> fields)
+        public void Visit(ITypeDefinition containingType, IEnumerable<IFieldDefinition> fields)
         {
-            foreach (IFieldDefinition field in fields)
+            if (containingType.IsValueType)
```

Comment:
I think we should call Visit only for fields that stay and shouldn't call visit for fields that are excluded.